### ISSUE TITLE
Enhancements to reduce risks and increase profits 

### ISSUE THREAD

**Issue by Maxoos** [chrisleekr/binance-trading-bot#18]
First of all I have to say I love this little thing!

I have a few ideas to help reduce risk and increase profits:

1. Place stop-limit orders for buying too, the same way it is done in chaseStopLossLimitOrder. Could probably reuse that function to do both buy and sell.
2. Seperate params per symbol to manage risk and strategy (maybe even have presets!). For example, BTCDOWN/USDT is more risky, I would like to look at 5m intervals and allow a lower maximum purchase amount just for that coin.
3. Enable/disable buying per symbol. Sometimes I just want to exit without buying anything new for one specific coin.

Any thoughts? I might have a few hours this week to help out.

**Comment by OOtta** [comment]
I would also like the option to replace the stop-limit orders immediately, instead of waiting for next time the symbol is processed again.

**Comment by chrisleekr** [comment]
@Maxoos that is an amazing idea. I might need to come up with good logic of chase stop-loss limit order for buying. 

But yeah, I will try to find out the way.

I will develop the following changes based on your suggestions.

- [ ] Allow entering more decimals for the last buy price
- [ ] Override buy/sell configuration per symbol
- [ ] Use chase-stop-loss-limit order for buy signal as well
- [ ] Enable/Disable symbols transaction, but continue to monitor
- [ ] Override the lowest value in the frontend

@OOtta I am not sure I understand what you mean. Can you elaborate on what you are trying to achieve?

**Comment by OOtta** [comment]
@chrisleekr 
If there is an opened Stop-Limit order and price goes higher it canceles the order and sets new order next time it checks that ticker, however if the market is volatile the price might dip down before new order is set and therefore you end up missing possible stop trigger and sell. If you really believe the price is gonna go up its okay, because there will be another opurtunity in the future, however i would prefer to have the option to play it safer and from the moment minimum profit price is reached have always stop-limit order in that would just get updated if price goes up.

**Comment by kaziqta** [comment]
Yeah, experienced the same these days, we all did, I think, this week's dip wasn't pleasant :)
But still, there may be also extension to monitor the trend, EMA 200 does a great job to read the trend.

**Comment by Maxoos** [comment]
You can play with the stop and limit parameters to create a bigger buffer
zone but less potential profit.

**Comment by bavinsCo** [comment]
Guys, join the Telegram group to solve problems with each other ..

https://t.me/joinchat/z3cKyYFcx0NiOTA8

**Comment by chrisleekr** [comment]
@Maxoos 

I have updated few features. Although in this release, I haven't applied stop chase limit order for buying. Need some more time for that.

**Comment by Maxoos** [comment]
@chrisleekr 
AMAZING! thank you!

I've made a couple of changes on my local to improve the performance, I simply cannot find the time to post it a PR, can you post them here and you'll consider them?

**Comment by chrisleekr** [comment]
@Maxoos Definitely. I will update if you give me what the changes are.

**Comment by Maxoos** [comment]
@chrisleekr 
1. This seems to be a bug
https://github.com/chrisleekr/binance-trading-bot/blob/025be0af537d83ebf56a916f710d2be9783879ca/app/jobs/simpleStopChaser.js#L83
If the action is 'buy', it will never try to sell even if it's time to sell. You can have buy and sell at the same time if you changes parameters or entered last buy price manually. I think it should simply look for sell order all the time. I took this line outside the condition.

2. When there's a sharp drop in price, this condition does not work well:
https://github.com/chrisleekr/binance-trading-bot/blob/025be0af537d83ebf56a916f710d2be9783879ca/app/jobs/simpleStopChaser/helper.js#L702
What I have found is if the last candle is rarely lowestClosed, so under this condition you must wait for the next candle to be lower or equal to the last candle, because the min will never be lowestClosed. I've added some padding around this like so:

```
  const diff = ((1 - lastCandle.close / lowestClosed) * -100)

  logger.info(
    { symbol, lowestClosed, close: lastCandle.close, diff },
    "Diff indicators"
  );
  if (lastCandle.close <= lowestClosed || diff <= 1) {
/// do something 
}
```

Ideally you can put the padding in a param to be edited, but 1% seems to work nicely. With this, if in the array of candles the next candle is slightly higher than the min, it will signal 'buy'. I monitored the market it happens all the time, missing those orders. Like in this example:
![image](https://user-images.githubusercontent.com/1913291/110198894-bd928100-7ea9-11eb-9285-9c67b1d1ee90.png)

Not sure if my solution is the best, feel free to handle it as you think.


3. Now when I started to play with this, I placed low orders, around $12. When the value of the coin went down enough, it deleted the last buy price, so I just committed out the two places where you delete the last buy price. Not sure how you solve this one, but I assume it does not happen if you place a $100 order. Unless that coin loses a lot of value.
https://github.com/chrisleekr/binance-trading-bot/blob/025be0af537d83ebf56a916f710d2be9783879ca/app/jobs/simpleStopChaser/helper.js#L330

**Comment by chrisleekr** [comment]
1. Handling actions simultaneously

    Ok, I think it's now possible to handle buy/chase-stop-loss-limit-order at a single process.
    In the original concept, the last buy price was configured by the bot and didn't allow to edit manually. Hence, if the buy action has occurred, technically sell action should never be occurred because the minimum selling price will never be met.  
    I will update the code to handle both steps in a single process.

2. Buy price buffer

    Ok, I think I can introduce a new parameter `Buy Trigger Percentage`. If you configure `1.01` then the trigger price will be 1% over the lowest price. It will give you some padding for the lowest price.


3. Under $10

    Yes, that is not a bug, it is intended. The reason it removes the last buy price under $10 is that Binance does not allow to place an order under $10 - https://www.binance.com/en-AU/trade-rule. 
    Normally, after you sold your coin, it leaves some residue amount of coins, normally under $1. If I don't remove the last buy price, it will continuously monitor chase-stop-loss-limit.

**Comment by Maxoos** [comment]
@chrisleekr Amazing! thank you for doing this.

**Comment by chrisleekr** [comment]
@Maxoos I am doing this because I also learn from doing it. ;)

Anyway, just note that updated the bot with the following features.

- Add max-size for logging
- Execute chaseStopLossLimitOrder on every process
- Support buy trigger percentage

**Comment by Maxoos** [comment]
@chrisleekr AMAZING!!! thank you so much. Can't wait to try it. I'm gonna try the chaseStopLossLimitOrder on 5m candles because they go up and down all the time :)

**Comment by kaziqta** [comment]
> Hi,
> 
> When the market price reaches the min sell price, it creates STOP_LOSS_LIMIT with price lower than the original buy price. Can you please let me know when It actually sell?

You are using too low minimum profit %, would suggest you to read how Stop limit orders work in every market

**Comment by nithi2023** [comment]
@kaziqta I am using the following settings just to test it out: 

Minimum profit percentage : 1.01
Stop price percentage: 0.99
Limit price percentage: 0.985

I am just wondering why it is not selling when market reaches target sell price (1.01 x buy price). It creates STOP_LOSS_LIMIT order with price lower than original buy price.

**Comment by kaziqta** [comment]
> @kaziqta I am using the following settings just to test it out:
> 
> Minimum profit percentage : 1.01
> Stop price percentage: 0.99
> Limit price percentage: 0.985
> 
> I am just wondering why it is not selling when market reaches target sell price (1.01 x buy price). It creates STOP_LOSS_LIMIT order with price lower than original buy price.
> 
> I am looking to make this bot to buy possible dip and sell it when it reaches profit target. Let me know if I am wrong.

I just think that you just don't make effort to make the math. The bot is a good tool, you have to calculate.

Ok, following situation: a 100 position goes up by 1% and you set a sell order for the 99% stop price for the 99% of 101, which will trigger a sale at 98.5% (limit) of 101.

Question: at what price will you sell?

**Comment by nithi2023** [comment]
I quiet understand what is STOP price and LIMIT price.
What I thought was once profit % is reached it will create the sell order and make a profit. 
But, I think I confused this with OCO or I didn't read the help text well. 

Lets say, 
Buy Price : 50000
I would like to take only 1% profit. so, `Minimum profit percentage` will be `1.01` (50500).

and what should be the `Stop price percentage` & `Limit price percentage` ?

**Comment by chrisleekr** [comment]
@Maxoos 

The bot has been updated with trailing buy (stop-loss-limit-order for buying). 

https://github.com/chrisleekr/binance-trading-bot/releases/tag/v0.0.57

Make sure to go through README.md.

As I introduce a new feature, I did lots of refactoring the code including settings. If the bot version is lower than the version 0.0.57, then the update will cause lost your settings and the last buy price records. You must write down settings and the last buy price records and re-configure after the upgrade.

If experiences any issue, simply delete all docker volumes/images and re-launch the bot.

I close the issue as most features are delivered except `Override the lowest value in the frontend`. This is already on the list to do.
